week4/Q1.py

```diff
@@ -0,0 +1,104 @@
+def _rename_islands(islands_map, old_name, new_name):
+	'''
+	Changes names of the tiles.
+
+	Args:
+		islands_map: two-dimensional list
+		old_name: int
+		new_name: int
+
+	Returns:
+		two-dimensional list
+	'''
+	for i, row in enumerate(islands_map):
+		for j, tile in enumerate(row):
+			if tile == old_name:
+				islands_map[i][j] = new_name
+	return islands_map
+
+
+def _name_island(islands_map, row, column, islands, island_counter, upper_tile, left_tile):
+	'''
+	Gives name (which island) to the new tile, renames old tiles if needed (if islands which were seen as separate actually happen to be connected).
+
+	Args:
+		islands_map: two_dimensional list
+		row: int, row of the tile
+		column: int, column of the tile
+		islands: set, already existing islands
+		island_counter: int, last name given
+		upper_tile: int, value of the tile under which is new one
+		left_tile: int, value of the tile right to which is new one
+
+	Returns:
+		two-dimensional list, set, int
+	'''
+	if upper_tile:
+		islands_map[row][column] = upper_tile
+
+	if left_tile:
+		if islands_map[row][column] == 0:
+			islands_map[row][column] = left_tile
+		elif islands_map[row][column] != left_tile:
+			islands.discard(left_tile)
+			islands_map = _rename_islands(islands_map, left_tile, islands_map[row][column])
+
+	if islands_map[row][column] == 0:
+		island_counter += 1
+		islands_map[row][column] = island_counter
+		islands.add(island_counter)
+
+	return islands_map, islands, island_counter
```

Comment:
Arrays in Python are passed by reference.

That means that if a function receives an array and modifies it, the caller's array is also modified (there is no copy). Therefore, the function does not need to return the array itself. In fact, this is misleading, because one might assume that the original is left unmodified because one is returned (it might be a copy that has been modified while the original is returned).

For example:
```python
def reset_matrix(matrix, value=0):
    """Resets all the values stored in a two-dimensional matrix.

    Args:
      matrix: a list of lists, the two-dimensional matrix.
      value: the value to set all entries in the matrix to. Default is 0.
    """
    for row in range(len(matrix)):
        for col in range(len(matrix[row])):
            matrix[row][col] = value

matrix = [
        [1, 0, 0],
        [0, 1, 0],
        [0, 0, 1],
]

print('Original:')
print(matrix)
reset_matrix(matrix)
print('Reset:')
print(matrix)
reset_matrix(matrix, value=42)
print('Reset to 42:')
print(matrix)
```
prints:
```python
Original:
[[1, 0, 0], [0, 1, 0], [0, 0, 1]]
Reset:
[[0, 0, 0], [0, 0, 0], [0, 0, 0]]
Reset to 42:
[[42, 42, 42], [42, 42, 42], [42, 42, 42]]
```

This is reasonably common in Python.

For example, `list.sort()` modifies the original list and does not return anything:
```python
>>> l = [2, 3, 1]
>>> l.sort()
>>> l
[1, 2, 3]
```
while `string.upper()` does not modify the original string (because strings are immutable in Python) and therefore has to return the new string.
```python
>>> s = 'AbcD'
>>> s.upper()
'ABCD'
>>> s
'AbcD'
```

One important thing in writing good code is that it behaves in a way that is intuitive and consistent with people's expectations.